src/poly/multiopen/verifier.rs

```diff
@@ -0,0 +1,246 @@
+                );
+            }
+        }
+
+        // Sample a challenge x_5 for keeping the multi-point quotient
+        // polynomial terms linearly independent.
+        let x_5: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        // Obtain the commitment to the multi-point quotient polynomial f(X).
+        hash_point(transcript, &self.f_commitment).unwrap();
+
+        // Sample a challenge x_6 for checking that f(X) was committed to
+        // correctly.
+        let x_6: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        for eval in self.q_evals.iter() {
+            transcript_scalar.absorb(*eval);
+        }
+
+        let transcript_scalar_point =
+            C::Base::from_bytes(&(transcript_scalar.squeeze()).to_bytes()).unwrap();
+        transcript.absorb(transcript_scalar_point);
+
+        // lagrange_Interpolate polynomial for evaluations at each set
+        let mut r_evals = vec![C::Scalar::zero(); point_sets.len()];
+        let mut r_polys: Vec<Vec<C::Scalar>> = Vec::with_capacity(point_sets.len());
+        for (points, evals) in point_sets.clone().iter().zip(q_eval_sets.clone().iter()) {
+            r_polys.push(lagrange_interpolate(&points, &evals));
+        }
+        for (r_eval, r_poly) in r_evals.iter_mut().zip(r_polys.iter()) {
+            *r_eval = eval_polynomial(r_poly, x_6);
+        }
+
+        // We can compute the expected msm_eval at x_6 using the q_evals provided
+        // by the prover and from x_5
+        let mut msm_eval = C::Scalar::zero();
+        for (set_idx, points) in point_sets.iter().enumerate() {
+            let mut eval = self.q_evals[set_idx];
+            eval -= &r_evals[set_idx];
+            for point in points {
+                eval = eval * &(x_6 - &point).invert().unwrap();
+            }
+
+            msm_eval *= &x_5;
+            msm_eval += &eval;
+        }
+
+        // Sample a challenge x_7 that we will use to collapse the openings of
+        // the various remaining polynomials at x_6 together.
+        let x_7: C::Scalar = get_challenge_scalar(Challenge(transcript.squeeze().get_lower_128()));
+
+        // Compute the final commitment that has to be opened
+        let mut commitment_msm = params.empty_msm();
+        commitment_msm.add_term(C::Scalar::one(), self.f_commitment);
+        for (set_idx, _) in point_sets.iter().enumerate() {
+            commitment_msm.scale(x_7);
+            commitment_msm.add_msm(&q_commitments[set_idx]);
+            msm_eval *= &x_7;
+            msm_eval += &self.q_evals[set_idx];
+        }
+
+        // Verify the opening proof
+        self.opening
+            .verify(params, msm, transcript, x_6, commitment_msm, msm_eval)
+    }
+}
+
+// For multiopen verifier: Construct intermediate representations relating commitments to sets of points by index
+fn construct_intermediate_sets<'a, C: CurveAffine, I>(
```

Comment:
It seems like you could potentially unify the logic in this function with the logic in the `prover`'s version of this function using generics, but maybe it's needlessly complicated.